yellowbrick/utils/kneed.py

```diff
@@ -0,0 +1,208 @@
+            will detect elbows.
+        :type curve: string
+        :param direction: one of {"increasing", "decreasing"}
+        :type direction: string
+        """
+        # Step 0: Raw Input
+        self.x = x
+        self.y = y
+        self.curve = curve
+        self.direction = direction
+        self.N = len(self.x)
+        self.S = S
+
+        # Step 1: fit a smooth line
+        uspline = interpolate.interp1d(self.x, self.y)
+        self.Ds_x = np.linspace(np.min(self.x), np.max(self.x), self.N)
+        self.Ds_y = uspline(self.Ds_x)
+
+        # Step 2: normalize values
+        self.xsn = self.__normalize(self.Ds_x)
+        self.ysn = self.__normalize(self.Ds_y)
+
+        # Step 3: Calculate difference curve
+        self.xd = self.xsn
+        if self.curve == 'convex' and direction == 'decreasing':
+            self.yd = self.ysn + self.xsn
+            self.yd = 1 - self.yd
+        elif self.curve == 'concave' and direction == 'decreasing':
+            self.yd = self.ysn + self.xsn
+        elif self.curve == 'concave' and direction == 'increasing':
+            self.yd = self.ysn - self.xsn
+        if self.curve == 'convex' and direction == 'increasing':
+            self.yd = abs(self.ysn - self.xsn)
+
+        # Step 4: Identify local maxima/minima
+        # local maxima
+        self.xmx_idx = argrelextrema(self.yd, np.greater)[0]
+        self.xmx = self.xd[self.xmx_idx]
+        self.ymx = self.yd[self.xmx_idx]
+
+        # local minima
+        self.xmn_idx = argrelextrema(self.yd, np.less)[0]
+        self.xmn = self.xd[self.xmn_idx]
+        self.ymn = self.yd[self.xmn_idx]
+
+        # Step 5: Calculate thresholds
+        self.Tmx = self.__threshold(self.ymx)
+
+        # Step 6: find knee
+        self.knee, self.norm_knee, self.knee_x = self.find_knee()
+
+    @staticmethod
+    def __normalize(a):
+        """normalize an array
+        :param a: The array to normalize
+        :type a: array
+        """
+        return (a - min(a)) / (max(a) - min(a))
+
+    def __threshold(self, ymx_i):
+        """Calculates the difference threshold for a
+        given difference local maximum
+        :param ymx_i: the normalized y value of a local maximum
+        """
+        return ymx_i - (self.S * np.diff(self.xsn).mean())
+
+    def find_knee(self, ):
+        """This function finds and returns the knee value, the normalized knee
+        value, and the x value where the knee is located.
+        :returns: tuple(knee, norm_knee, knee_x)
+        :rtype: (float, float, int)
+        )
+        """
+        if not self.xmx_idx.size:
+            warnings.warn("No local maxima found in the distance curve\n"
+                          "The line is probably not polynomial, try plotting\n"
+                          "the distance curve with plt.plot(knee.xd, knee.yd)\n"
+                          "Also check that you aren't mistakenly setting the curve argument", RuntimeWarning)
```

Comment:
Unfortunately, I think we may have to do something about this warning, otherwise, it could be very confusing to our users. 

First, it is probably preferable to raise an exception here - that way we can catch the exception in the `KElbowVisualizer` and either keep drawing without the detected elbow, issue an elbow-specific warning, or advise the user to set `locate_elbow=False`. I think this warning will be issued given the convex/decreasing or concave/increasing issue that I mentioned above. So we may want to do things differently depending on if the user has the ability to control these params or something has gone wrong with what we expected for the metric. 

If we do go with a warning, instead of a RuntimeWarning, could we please issue a [`YellowbrickWarning `](https://github.com/DistrictDataLabs/yellowbrick/blob/develop/yellowbrick/exceptions.py#L99)?

Reply:
whew, this is very, very legacy... the warning started as a sanity check long ago.

Reply:
We are running into it because we can be a bit all over the place with convex/increasing or concave/decreasing depending on the metric we're using -- and if the clustering is terrible (e.g. bad features, wrong algorithm, no actual clusters) then things get really wild at that point. Any advice or thoughts you have would be welcome!

Reply:
@bbengfort I am trying to get familiar with the exceptions handling. Once done I'll accommodate these changes in a PR. 
Thanks.